File: Flink/src/main/scala/realtime/flink/scala/streaming/WordCount.scala
package realtime.flink.scala.streaming

import org.apache.flink.streaming.api.scala._
import org.apache.flink.util.Collector
import org.apache.flink.api.common.functions.{FlatMapFunction, RichFlatMapFunction}
import org.apache.flink.api.java.utils.ParameterTool;
/**
 *    本程序重点是了解Flink的编程结构以及Scala编程实现的简便之处
 *    执行需要添加分析文件信息，--input src\main\resources\WordCountData.txt
 */

class MyFlatMapFunction extends FlatMapFunction[String, String] {
  override def flatMap(t : String, out : Collector[String]) : Unit = {
    t.toLowerCase.split(" ").foreach(out.collect)
  }
}

class MyRichFlatMapFunction extends RichFlatMapFunction[String, String] {
  override def flatMap(in: String, out: Collector[String]): Unit = {
    in.toLowerCase.split(" ").foreach(out.collect)
  }
}

object WodCount {
  def main(args: Array[String]): Unit = {
    val inputParam = ParameterTool.fromArgs(args)
    if (!inputParam.has("input")) {
      System.out.println("Please make sure your input!")
      return;
    }

    // 1. 创建执行环境
    val env = StreamExecutionEnvironment.getExecutionEnvironment;
    // 2. 获取输入
    val text = env.readTextFile(inputParam.get("input"))

    // 3. 进行数据转换操作
    val counts: DataStream[(String, Int)] = text
      //      .flatMap(_.toLowerCase.split(" "))                // 实现1：FlapMap算子
      //       .flatMap(new MyFlatMapFunction)                  // 实现2：采用创建Class实现Funciton接口
      //        .flatMap(new FlatMapFunction[String, String] {
      //          override def flatMap(t : String, out : Collector[String]) : Unit = {
      //            t.toLowerCase.split(" ").foreach(out.collect)
      //          }
      //        })                                            // 实现3: 采用创建匿名类实现Funciton接口
      .flatMap(new MyRichFlatMapFunction)               // 实现4：通过RichFunciton实现接口
      .filter(_.nonEmpty)
      .map((_, 1))
      .keyBy(value => value._1)
      .sum(1)

    // 4. 指定计算结果输出位置
    if (inputParam.has("output")) {
      counts.writeAsText(inputParam.get("output"))
    } else {
      println("Printing result to stdout. Use --output to specify output path.")
      counts.print()
    }

    // 5. 指定名称并触发流式任务
    env.execute("Streaming WordCount")
  }
}

File: Flink/src/main/java/realtime/flink/java/streaming/WordCount.java
package realtime.flink.java.streaming;

import geek.flink.base.WindowWordCount;
import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.api.java.utils.MultipleParameterTool;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;
import org.apache.flink.util.Preconditions;

public class WordCount {

    /**
     * 将每一行数据进行切分转换为(word, 1)的形式
     */
    public static class TextSpliter implements FlatMapFunction<String, Tuple2<String, Integer>> {
        public void flatMap(String sentence, Collector< Tuple2<String, Integer> > out) {
            for (String word : sentence.toLowerCase().split("\\W+")) {
                out.collect(new Tuple2<>(word, 1));
            }
        }
    }

    public  static void main(String[] args) {
        try {
            // 读取输入参数
            MultipleParameterTool inputParams = MultipleParameterTool.fromArgs(args);
            if (!inputParams.has("input")) {
                System.out.println("Please make your input!");
                return;
            }

            // 设置运行环境
            StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
            // 使Web接口也使用该参数
            env.getConfig().setGlobalJobParameters(inputParams);
            // 获取数据源并进行一些转换操作
            DataStream<String> text = null;
            for (String in : inputParams.getMultiParameter("input")) {
                if (text == null) {
                    text = env.readTextFile(in);
                } else {
                    text.union(env.readTextFile(in));
                }
            }
            Preconditions.checkNotNull(text, "Input DataStream should not be null.");
            DataStream<Tuple2<String, Integer>> dataStream = text
                    .flatMap(new WindowWordCount.TextSpliter())
                    // 根据Tuple中的0字段进行分组并将第1个字段相加
                    .keyBy(value -> value.f0)
                    .sum(1);

            // Sink操作
            if (inputParams.has("output")) {
                dataStream.writeAsText(inputParams.get("output"));
            } else {
                dataStream.print();
            }
            // 执行应用程序
            env.execute("Flink Stream WordCount");
        } catch (Exception e) {
            e.printStackTrace();
            System.out.println("Meet exception: " + e.getMessage());
        }
    }
}
